pkg/operator/status.go

```diff
@@ -289,9 +291,99 @@ func (optr *Operator) syncUpgradeableStatus() error {
 		coStatus.Message = "One or more machine config pools are updating, please see `oc get mcp` for further details"
 	}
 
+	// don't overwrite status if updating and degraded
+	if !updating && !degraded {
+		if supported, v := optr.isKubeletSkewSupported(); !supported {
+			coStatus.Reason = "KubeletSkewUnsupported"
+			coStatus.Message = fmt.Sprintf("machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least %v.", getMinimalSkewSupportNodeVersion(v))
+		}
+	}
+
 	return optr.updateStatus(co, coStatus)
 }
 
+// isKubeletSkewSupported checks the version skew of kube-apiserver and node kubelet version.
+// Returns false is the version skew is not supported. version skew >= 2 is not supported.
+func (optr *Operator) isKubeletSkewSupported() (skewSupported bool, kubeAPIServerVersion string) {
+	skewSupported = true
+	nodes, err := optr.nodeLister.List(labels.Everything())
+	if err != nil {
+		return skewSupported, ""
+	}
+	kubeAPIServerStatus, err := optr.configClient.ConfigV1().ClusterOperators().Get(context.TODO(), "kube-apiserver", metav1.GetOptions{})
+	if err != nil {
+		return skewSupported, ""
+	}
+
+	// looks like
+	//   - name: kube-apiserver
+	//     version: 1.21.0-rc.0
+	kubeAPIServerVersion = ""
+	for _, version := range kubeAPIServerStatus.Status.Versions {
+		if version.Name != "kube-apiserver" {
+			continue
+		}
+		kubeAPIServerVersion = version.Version
+		break
+	}
+	if kubeAPIServerVersion == "" {
+		return skewSupported, kubeAPIServerVersion
+	}
+	kubeAPIServerMinorVersion, err := getMinorKubeletVersion(kubeAPIServerVersion)
+	if err != nil {
+		return skewSupported, kubeAPIServerVersion
+	}
+
+	for _, node := range nodes {
+		// looks like kubeletVersion: v1.21.0-rc.0+6143dea
+		kubeletVersion := node.Status.NodeInfo.KubeletVersion
+		if kubeletVersion == "" {
+			continue
+		}
+		nodeMinorVersion, err := getMinorKubeletVersion(kubeletVersion)
+		if err != nil {
+			glog.Error(err)
+			continue
+		}
+		if nodeMinorVersion+2 <= kubeAPIServerMinorVersion {
```

Comment:
note: we're checking on this

Reply:
@kikisdeliveryservice  I checked this with @rphillips . We can keep it as is. Considering the 2nd example here https://docs.google.com/document/d/1yXzWPBWmKLuKBkvOKMk0FusyBpUqjyT3hGXYL8tUafU/edit#. we get into this situation when we do upgrade kube-apiserver is at 1.21 and 1.20 , then the kubelet can only support 1.20 and 1.19.

Reply:
Wouldn't <= mean that we only support 1? if apiserver is at 1.20 and node is at 1.18 and I add 2 it's = apiserver and not supported. Is this correct?

Reply:
Up to n-2 is support.

1. 1.21 kas supports 1.21, 1.20, 1.19 kubelet
2. 1.20 kas supports 1.20, 1.19, 1.18 kubelet

The allowable skew is the intersection, in this case 1.20 and 1.19.

Reply:
> Up to n-2 is support.
> 
>     1. 1.21 kas supports 1.21, 1.20, 1.19 kubelet
> 
>     2. 1.20 kas supports 1.20, 1.19, 1.18 kubelet
> 
> 
> The allowable skew is the intersection, in this case 1.20 and 1.19.

@deads2k So the skew calculation should take into account the upgrade, like both 1.21 and 1.20 in the above example?

Reply:
right but my question is that <= correct?

and isn't apiserver has various versions only applicable when we're mid-update? post update, they would be the same.